plugin: Support LogOutputChannel

ndoschek opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See SECURITY.md at the root of this repository, to learn how to report
vulnerabilities.
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->

- Support LogOutputChannel
- Support LogLevel
- Support in namespace/env: logLevel & onDidChangeLogLevel

Resolves #12017

Contributed on behalf of STMicroelectronics.

##### Remark:
This PR contributes the basic VS Code API support for LogOutputChannels, to fully support all aspects of a LogOutputViewChannel as in VS Code some additional features will be needed:
  - In VSCode there is support to set the "Developer log level" in the application which also configures the log level for LogOutputviewChannels. Currently the log level for LogOutputChannels is set to the default log level, i.e. info.
  - VSCode also writes the logs to a separate file per extension and offers a command and tool in the output view to directly open the extensions's log file.
 
I suggest to extract those additional features into separate follow-up issues if you agree.

[Edit 2023-04-21]: Extracted to follow-up issue #12440

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

To test, I created a simple extension that creates a LogOutputChannel and contributes commands to send log messages in different levels:
[Edit 2023-04-21: updated VSIX test extension] [vscode-extension-log-output-channel-12017-0.0.1.zip](https://github.com/eclipse-theia/theia/files/11294198/vscode-extension-log-output-channel-12017-0.0.1.zip)



1. Download and install the test extension `vscode-extension-log-output-channel-12017-0.0.1.vsix` in your Theia test workspace (via `Extensions` view - top menu bar `More actions...` - `Install from VSIX`).
2. On activation of the extension, the log output channel `LogOutputChannel-Test-12017` is shown in the output view and the initial message (`LogLevel set to: Info`) should be shown.
3. Execute the commands (`Log a <loglevel> message via LogOutputChannel (with args)` / `Throw an ERROR and log msg/stack via LogOutputChannel` / `Test replace LogOutputChannel` / `Append text content (as new line) to LogOutputChannel`) via the command palette and check for the messages in the output channel.
Remark: Currently the loglevel is set to default INFO log level.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

ndoschek commented:
We will do a first internal review cycle and mark this one as ready for review afterwards.

kittaakos commented on packages/plugin/src/theia.d.ts:
```diff
@@ -5250,6 +5353,15 @@ export module '@theia/plugin' {
 
+        /**
+         * Creates a new {@link OutputChannel output channel} with the given name.
+         * If options are given, creates a new {@link OutputChannel output channel} with the given name.
```
This does not make much sense here. `options` is mandatory, and it will create a `LogOutputChannel` instance and not an `OutputChannel`. What about copying the docs from Code:

> Creates a new [log output channel](https://code.visualstudio.com/api/references/vscode-api#LogOutputChannel) with the given name.

kittaakos commented on packages/plugin-ext/src/common/plugin-api-rpc.ts:
```diff
@@ -286,7 +286,8 @@ export interface TerminalServiceExt {
 export interface OutputChannelRegistryExt {
-    createOutputChannel(name: string, pluginInfo: PluginInfo): theia.OutputChannel
+    createOutputChannel(name: string, pluginInfo: PluginInfo): theia.OutputChannel,
+    createOutputChannel(name: string, pluginInfo: PluginInfo, options?: { log: true }): theia.LogOutputChannel
```
`options` is not optional here.

tsmaeder replied:
This is an internal interface, so we can make it whatever we want ;-). But you're right, it probably shouldn't be.

kittaakos reviewed (commented):
It works great 👍  I even verified with an `Error` after modifying your test VSIX:


```
2023-04-20T07:52:54.937Z [Error] Error: Logging an ERROR message via LogOutputChannel
    at /Users/a.kitta/dev/git/theia/plugins/vscode-extension-log-output-channel-12017-0.0.1/extension/out/extension.js:40:34
    at /Users/a.kitta/dev/git/theia/packages/plugin-ext/lib/plugin/command-registry.js:61:59
    at CommandRegistryImpl.executeLocalCommand (/Users/a.kitta/dev/git/theia/packages/plugin-ext/lib/plugin/command-registry.js:108:20)
    at CommandRegistryImpl.$executeCommand (/Users/a.kitta/dev/git/theia/packages/plugin-ext/lib/plugin/command-registry.js:70:25)
    at /Users/a.kitta/dev/git/theia/packages/plugin-ext/lib/common/proxy-handler.js:91:71
    at process.processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async RpcProtocol.handleRequest (/Users/a.kitta/dev/git/theia/packages/core/lib/common/message-rpc/rpc-protocol.js:167:28)
```

Regarding the API, when creating the `LogOutputChannel`, `options` must be defined, and `log` must be `true` to be in sync with Code API. What do you think?

tsmaeder commented on packages/plugin/src/theia-proposed.d.ts:
```diff
@@ -168,35 +168,6 @@ export module '@theia/plugin' {
-     * The severity level of a log message
-     */
-    export enum LogLevel {
-        Trace = 1,
```
It's only "proposed", but we might want to add a breaking change notice in case an adopter is using this API.

tsmaeder commented on packages/plugin-ext/src/plugin/output-channel-registry.ts:
```diff
@@ -13,27 +13,39 @@
-        } else {
-            return new OutputChannelImpl(name, this.proxy, pluginInfo);
         }
+        const isLogOutput = options && typeof options === 'object' && options.log;
```
There is also `isObject<T>(...)` from types.ts

tsmaeder commented on packages/plugin-ext/src/plugin/output-channel/logoutput-channel.ts:
```diff
@@ -0,0 +1,80 @@
+import { OutputChannelRegistryMain, PluginInfo } from '../../common/plugin-api-rpc';
+import { toLogLevel } from '../type-converters';
+
+export class LogOutputChannelImpl extends AbstractMessageLogger implements theia.LogOutputChannel {
```
I believe the proper filename according to conventions would be `log-output-channel.ts` or else it would have to be `LogoutputChannel`.

ndoschek replied:
Thanks for raising this Thomas! I was a bit biased by vscode's implementation I guess.
I had a look and it makes a lot of sense to extend the `OutputChannelImpl`. However, I mimicked a bit of logic from the formerly used Logger. I marked a block of code as a copy, could you have another look if this is okay this way? Thanks!

tsmaeder commented on packages/plugin-ext/src/plugin/output-channel/logoutput-channel.ts:
```diff
@@ -0,0 +1,80 @@
+
+    override onDidChangeLogLevel: theia.Event<theia.LogLevel>;
+
+    constructor(readonly name: string, protected proxy: OutputChannelRegistryMain, protected readonly pluginInfo: PluginInfo) {
```
is the `proxy` not readonly?

ndoschek replied:
Good catch, thank you, I agree and also think it should be readonly. However, it wasn't readonly in the `OutputChannelImpl` either, hence I updated it for both channels.

tsmaeder commented on packages/plugin-ext/src/plugin/type-converters.ts:
```diff
@@ -1385,3 +1386,16 @@ export namespace DataTransfer {
+
+export function toLogLevel(logLevel: MonacoLogLevel): theia.LogLevel {
+    switch (logLevel) {
+        case MonacoLogLevel.Trace: return types.LogLevel.Trace;
```
One more argument for using `OutputChannelImpl` as a base class: we wouldn't need this conversion.

tsmaeder reviewed (changes requested):
I don't think using the logger implementation from VS Code is a good idea in this case. Feel free to convince me, though :-)

ndoschek commented on packages/plugin-ext/src/plugin/output-channel/logoutput-channel.ts:
```diff
@@ -0,0 +1,80 @@
+    replace(value: string): void {
+        this.info(value);
+        this.proxy.$append(this.name, value, this.pluginInfo);
+    }
```
Thanks for catching that! It basically should replace the channel's content with the given text value.
With the updated implementation, we reuse the `OutputChannelImpl` and can reuse the logic, which clears the content first and then appends the given text value.
Also, I updated the test extension, which also offers a command now to call the replace function.

ndoschek commented:
Thank you both for your reviews!
I updated my implementation and also updated the test extension in the PR description (it now offers a few more commands like replace and error throwing)

Regarding the additional log file functionality in VS Code (see my remark in the PR description), is it okay if I add a follow-up for this?

tsmaeder commented on packages/plugin-ext/src/plugin/output-channel/log-output-channel.ts:
```diff
@@ -0,0 +1,132 @@
+        }
+    }
+
+    private format(args: any): string {
```
Isn't this just doing this?
```
args.map(arg => stringify the arg here).join(' ');
```

tsmaeder commented on packages/plugin-ext/src/plugin/output-channel/output-channel-item.ts:
```diff
@@ -20,7 +20,7 @@ export class OutputChannelImpl implements theia.OutputChannel {
     private disposed: boolean;
 
-    constructor(readonly name: string, private proxy: OutputChannelRegistryMain, private readonly pluginInfo: PluginInfo) {
+    constructor(readonly name: string, protected readonly proxy: OutputChannelRegistryMain, protected readonly pluginInfo: PluginInfo) {
```
This is a breaking change, as well, I think.

ndoschek commented on packages/plugin-ext/src/plugin/output-channel-registry.ts:
```diff
@@ -13,27 +13,40 @@
 export class OutputChannelRegistryExtImpl implements OutputChannelRegistryExt {
 
-    proxy: OutputChannelRegistryMain;
+    private proxy: OutputChannelRegistryMain;
```
I'm not sure why I changed that to private in the first commit, I reverted that change again.

ndoschek commented on packages/plugin-ext/src/plugin/output-channel/log-output-channel.ts:
```diff
@@ -0,0 +1,132 @@
+    readonly onDidChangeLogLevel: theia.Event<theia.LogLevel> = this.onDidChangeLogLevelEmitter.event;
+    public logLevel: theia.LogLevel;
+
+    constructor(override readonly name: string, protected override readonly proxy: OutputChannelRegistryMain, protected override readonly pluginInfo: PluginInfo) {
```
It is also readonly in the superclass, so I guess this is okay this way

tsmaeder replied:
But why would you redeclare a variable that is already present in the superclass and visible just fine in the subclass? It's unnecessary for all three constructor params, IMO.

ndoschek replied:
Yes basically I agree, but as I had to write a custom constructor (to initalize the log level) I have no other choice, as Typescript requires them to be overridden due to the same variable names, see `This parameter property must have an 'override' modifier because it overrides a member in base class 'OutputChannelImpl'.ts(4115)`

tsmaeder replied:
This works for me:
```
constructor(name: string, proxy: OutputChannelRegistryMain, pluginInfo: PluginInfo) {
        super(name, proxy, pluginInfo);
        this.setLogLevel(LogLevel.Info);
    }
```

The "readonly" is a property of the implicitly created field, not the constructor parameter.

ndoschek replied:
Thanks  I was not aware of that, that works great! I pushed another update for this.

ndoschek commented on packages/plugin-ext/src/plugin/output-channel/log-output-channel.ts:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    override append(value: string): void {
+        this.info(value);
```
Yes it does not explicitely say that, but if we do use the basic append mechanism, the output would look like this:

```
2023-04-21T08:13:28.630Z [Warning] Logging a WARNING message via LogOutputChannel
2023-04-21T08:13:33.981Z [Error] Logging an ERROR message via LogOutputChannel
Append text content as new line to LogOutputChannel
2023-04-21T08:13:41.973Z [Warning] Logging a WARNING message via LogOutputChannel
Append text content to LogOutputChannel2023-04-21T08:13:49.676Z [Warning] Logging a WARNING message via LogOutputChannel
```

Using `.info()` ensures that all logged output is formatted consistently.
VS Code has implemented it the same way, which I would prefer as well as the log output is overall more consistent.
But if you insist, we can also change it to the simple append mechanism.

tsmaeder replied:
Hmh...that's odd, why have `append` if it's exactly like `info`? But since this is VS Code API, compatibility trumps common sense. I'd leave it like it is.

ndoschek commented on packages/plugin-ext/src/plugin/output-channel/log-output-channel.ts:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    trace(message: string, ...args: any[]): void {
+        if (this.checkLogLevel(LogLevel.Trace)) {
```
That's true, thanks for catching that! I changed it to your suggestion.

tsmaeder commented on packages/plugin-ext/src/plugin/output-channel/output-channel-item.ts:
```diff
@@ -65,7 +65,7 @@ export class OutputChannelImpl implements theia.OutputChannel {
     }
 
-    private validate(): void {
+    protected validate(): void {
```
This one is breaking, too, as a subclass might already have this method.

ndoschek replied:
Yes thanks, I already added it to the Changelog breaking changes with the last commit.

tsmaeder reviewed (commented):
Some nitpicks left.

tsmaeder commented:
@ndoschek I just have two nitpicks left, then we're good to go. Works as advertised.

ndoschek commented:
Thank you @tsmaeder for your review! I left my replies in the open threads.

tsmaeder reviewed (approved):
Looks good to me, now.
